server/blob/blob-api/src/main/java/org/apache/james/blob/api/Store.java

```diff
@@ -0,0 +1,114 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.util.FluentFutureStream;
+
+import com.google.common.collect.ImmutableMap;
+
+public interface Store<T> {
```

Comment:
I see we have another `BlobStore` class, what is the difference between `BlobStore` and `Store`? And the names make me thinking `BlobStore` is a child of `Store`

Reply:
Ok, then...

> what is the difference between BlobStore and Store? A

BlobStore is a class responsible for storing (potentially large) chunk of bytes.

Store is initially designed to allow storing arbitrary objects in a blob store (but might evolve in storing arbitrary objects). Note that BlobStore is a low-level primitive (storing bytes), while Store also handles the following questions:

 - How to extract the bytes from my object
 - Do I need to cut it in parts before storing it?

So Store is actually very different from BlobStore...

> the names make me thinking BlobStore is a child of Store

Well, no big deal IMO a BlobStore storing blobs while a Store can store arbitrary things... The thing is that our arbitrary store have to depend on a specific store. I don't see any conceptual problem in this. And this should not push us, by trying to give names that would highlight the hierarchy difference you care about, to hide what these classes are really doing: storing stuff.

Reply:
I don't think it's useful: if you look at it, many interface consists just in read() and save() methods. I don't think we need to generalize them yet.

Reply:
I like the design but as raphael said, it probably won't compile because you don't have specialization in Java